### DISCUSSION TITLE
RFC: expose next build outputs programatically

### DISCUSSION THREAD

**Discussion by cdaringe** [vercel/next.js#16483] [Ideas]
# Problem

Next emits tons of great information, but much of that information is not available to users outside of stdio. Consider the output of `yarn build`:

![image](https://user-images.githubusercontent.com/1003261/90968340-0a7bb080-e4a0-11ea-8ac1-d3b436b1935e.png)

As seen above, you can see that next.js finds it reasonable to share a bunch essential build information to users via stdout. Similarly, I have questions, reports, and rules/enforcements that I want to generate or apply, but need much of that information programmatically. I assume all parties agree parsing stdout is not a practical option to access this type of information.  Some practical usages that this information would be useful for:

- Assert total first load JS to be less than X kB
- Forbid too many or too few chunks
- Assert SSG/Server/Static for pages X, Y, or Z for my app
- Assert X kB threshold for CSS
- Add #-pages, sum(count-assets), sum(count-js-kB) to trending reports

tl;dr, next has great internal stats--I want access to them!

# Solution

Add an API to tap into these entities. #16418 is a simple-stupid approach to enable this, and more.

**Comment by llwt** [comment]
> This would break in every single release as these properties are not for public use. Making a change like this would require a proper RFC.

@timneutkens,

If a subset of the build-stats data was passed into the proposed `onBuildComplete` would that help reduce potential breakages across releases? Maybe just `pageInfos` to start?

**Comment by freddydumont** [comment]
I'd use this in a GitHub action on PRs, to compare stats against the base branch and push a comment with the info.

I'm pretty sure I saw something like this before, but maybe it wasn't a Next.js project.

**Reply by llwt** [reply]
The issue is getting access to the report info.

Are you proposing we parse the `stdout` from github actions?

**Reply by freddydumont** [reply]
I don't have a technical solution in mind. I was just presenting an extra use case for this RFC.

**Reply by llwt** [reply]
Ahh gotcha, my bad misreading that haha 🤦

**Comment by jnv** [comment]
Perhaps another possible use case would be for creating a sitemap at build time.
I also wonder whether it would make sense to expose such functionality through the proposed plugin API: #9133

**Comment by cdaringe** [comment]
Here is another use case we have found recently. We build our app periodically in _different_ environments. locally? OSX. in CI? centos. for our prod builds? alpine.

We've found that build size outputs are _not consistent_ between build envs. Having programmatic output here would greatly help us root cause what files are different between build systems, s.t. we can improve consistency in our environments.

**Comment by conanbatt** [comment]
I would definitely like to have this feature, and even expand some of these stats.
My main use-case is appending bundle-size tracking to the CI process. Currently have a custom solution.

**Comment by max-pod** [comment]
Has there been any progress on exposing next build outputs?
